src/System.Runtime.Extensions/src/System/IO/PathHelper.cs

```diff
@@ -0,0 +1,459 @@
+            Contract.Requires(length == Path.MaxPath);
+
+            this.m_arrayPtr = charArrayPtr;
+            this.m_capacity = length;
+            this.m_maxPath = Path.MaxPath;
+            useStackAlloc = true;
+            doNotTryExpandShortFileName = false;
+        }
+
+        // Instantiates a PathHelper with a heap alloc'd array of ints. Will create a StringBuilder
+        internal PathHelper(int capacity, int maxPath)
+        {
+            this.m_sb = new StringBuilder(capacity);
+            this.m_capacity = capacity;
+            this.m_maxPath = maxPath;
+            doNotTryExpandShortFileName = false;
+        }
+
+        internal int Length
+        {
+            get
+            {
+                if (useStackAlloc)
+                {
+                    return m_length;
+                }
+                else
+                {
+                    return m_sb.Length;
+                }
+            }
+            set
+            {
+                if (useStackAlloc)
+                {
+                    m_length = value;
+                }
+                else
+                {
+                    m_sb.Length = value;
+                }
+            }
+        }
+
+        internal int Capacity
+        {
+            get
+            {
+                return m_capacity;
+            }
+        }
+
+        internal char this[int index]
+        {
+            [System.Security.SecurityCritical]
+            get
+            {
+                Contract.Requires(index >= 0 && index < Length);
+                if (useStackAlloc)
+                {
+                    return m_arrayPtr[index];
+                }
+                else
+                {
+                    return m_sb[index];
+                }
+            }
+            [System.Security.SecurityCritical]
+            set
+            {
+                Contract.Requires(index >= 0 && index < Length);
+                if (useStackAlloc)
+                {
+                    m_arrayPtr[index] = value;
+                }
+                else
+                {
+                    m_sb[index] = value;
+                }
+            }
+        }
+
+        [System.Security.SecurityCritical]
+        internal unsafe void Append(char value)
+        {
+            if (Length + 1 >= m_capacity)
+                throw new PathTooLongException(SR.IO_PathTooLong);
+
+            if (useStackAlloc)
+            {
+                m_arrayPtr[Length] = value;
+                m_length++;
+            }
+            else
+            {
+                m_sb.Append(value);
+            }
+        }
+
+        [System.Security.SecurityCritical]
+        internal unsafe int GetFullPathName()
+        {
+            if (useStackAlloc)
+            {
+                char* finalBuffer = stackalloc char[Path.MaxPath + 1];
+                int result = Interop.mincore.GetFullPathNameW(m_arrayPtr, Path.MaxPath + 1, finalBuffer, IntPtr.Zero);
```

Comment:
Same comment, we aren't sure why this is here, but we aren't building it into the CoreCLR libraries, so we are fine for today.